beacon-chain/endtoend/evaluators.go

```diff
@@ -0,0 +1,144 @@
+			t.Run(evaluator.name, func(t *testing.T) {
+				if err := evaluator.evaluation(client); err != nil {
+					t.Fatal(err)
+				}
+			})
+		}
+	}
+}
+
+func afterNEpochs(epochs uint64) func(uint64) bool {
+	return func(currentEpoch uint64) bool {
+		return currentEpoch == epochs
+	}
+}
+
+func onGenesisEpoch(currentEpoch uint64) bool {
+	return currentEpoch == 0
+}
+
+// validatorsActivate ensures the expected amount of validators
+// are active.
+func validatorsActivate(client eth.BeaconChainClient) error {
+	validatorRequest := &eth.GetValidatorsRequest{}
+	validators, err := client.GetValidators(context.Background(), validatorRequest)
+	if err != nil {
+		return fmt.Errorf("failed to get validators: %v", err)
+	}
+
+	expectedCount := params.BeaconConfig().MinGenesisActiveValidatorCount
+	receivedCount := uint64(len(validators.Validators))
+	if expectedCount != receivedCount {
+		return fmt.Errorf("expected validator count to be %d, recevied %d", expectedCount, receivedCount)
+	}
+
+	for _, val := range validators.Validators {
+		if val.ActivationEpoch != 0 {
+			return fmt.Errorf("genesis validator epoch should be 0, received %d", val.ActivationEpoch)
+		}
+		if val.WithdrawableEpoch != params.BeaconConfig().FarFutureEpoch {
+			return fmt.Errorf("genesis validator withdrawable epoch should be far future, received %d", val.WithdrawableEpoch)
+		}
+	}
+	return nil
+}
+
+// validatorsParticipating ensures the validators have an acceptable participation rate.
+func validatorsParticipating(client eth.BeaconChainClient) error {
+	in := new(ptypes.Empty)
+	chainHead, err := client.GetChainHead(context.Background(), in)
+	if err != nil {
+		return fmt.Errorf("failed to get chain head: %v", err)
+	}
+	currentEpoch := chainHead.BlockSlot / params.BeaconConfig().SlotsPerEpoch
+	if currentEpoch > 0 {
+		return fmt.Errorf("current epoch must be greater than 0, received %d", currentEpoch)
+	}
+
+	validatorRequest := &eth.GetValidatorParticipationRequest{
+		QueryFilter: &eth.GetValidatorParticipationRequest_Epoch{
+			Epoch: currentEpoch - 1,
+		},
+	}
+	participation, err := client.GetValidatorParticipation(context.Background(), validatorRequest)
+	if err != nil {
+		return fmt.Errorf("failed to get validator participation: %v", err)
+	}
+
+	partRate := participation.Participation.GlobalParticipationRate
+	if partRate < 0.85 {
+		return fmt.Errorf("validator participation not as high as expected, received: %f", partRate)
+	}
+	return nil
+}
+
+// finalizationOccurs is an evaluator to make sure finalization is performing as it should.
+// Requires to be run after at least 4 epochs have passed.
+func finalizationOccurs(client eth.BeaconChainClient) error {
+	in := new(ptypes.Empty)
+	chainHead, err := client.GetChainHead(context.Background(), in)
+	if err != nil {
+		return fmt.Errorf("failed to get chain head: %v", err)
+	}
+
+	currentEpoch := chainHead.BlockSlot / params.BeaconConfig().SlotsPerEpoch
+	if currentEpoch < 4 {
```

Comment:
should the magic numbers 4 and 2 be arguments? to make this more extensible and customizable later on